### PR TITLE
Allow overriding oauth callback url

### PR DESCRIPTION
**Pull request by paphonb**
When developing frontend locally using the dev backend, we want to use the proxied url for oauth callback in order to get authentication cookies.

See also: 
- thinc-org/cugetreg-frontend#271

### CODE REVIEW

File: src/auth/auth.controller.ts [hunk 1]
```diff
@@ -89,4 +90,29 @@ export class AuthController {
...
+    if (state.overrideBackendUrl) {
+      throw new BadRequestException(
+        `overriding backendUrl is not allowed in this environment`
+      )
+    }
+    if (!state.returnUrl.startsWith(backendUrl)) {
+      throw new BadRequestException(
+        `returnUrl must be the same origin as ${backendUrl}`
+      )
+    }
```
Comment (saengowp) [inline src/auth/auth.controller.ts:116]: Oh, This is not the correct way to check for origin isn't it?

**Review by saengowp — APPROVED** [review]
Nice.

File: src/auth/auth.controller.ts (hunk 1 above)
Comment (paphonb) [reply src/auth/auth.controller.ts:116]: Why not? Our `backendUrl` is an origin

Comment (saengowp) [reply src/auth/auth.controller.ts:116]: In https://github.com/thinc-org/cugetreg-infrastructure/blob/master/src/production/backend/.env.production, I could just do `https://cugetreg.com.malicious.domain/...` to bypass this check.

Comment (bombnp) [reply src/auth/auth.controller.ts:116]: how do u propose we check for origin then?

File: src/auth/auth.controller.ts [hunk 2]
```diff
@@ -89,4 +90,29 @@ export class AuthController {
...
     }
   }
+
+  private createOauthState(returnUrl: string, overrideBackendUrl?: string) {
+    const state: OauthStatePayload = {
+      returnUrl: returnUrl || this.configService.get('backendPublicUrl'),
+    }
+    if (overrideBackendUrl) {
+      state.overrideBackendUrl = overrideBackendUrl
+    }
```
Comment (bombnp) [inline src/auth/auth.controller.ts:100]: why not this?
```suggestion
    const state: OauthStatePayload = {
      returnUrl: returnUrl || this.configService.get('backendPublicUrl'),
      overrideBackendUrl: overrideBackendUrl
    }
```

**Review by bombnp — CHANGES_REQUESTED** [review]

File: src/auth/auth.controller.ts (hunk 2 above)
Comment (paphonb) [reply src/auth/auth.controller.ts:100]: Don't want that key in the object when it's undefined

Comment (bombnp) [reply src/auth/auth.controller.ts:100]: if it's undefined, it would be undefined, no?

Comment (paphonb) [reply src/auth/auth.controller.ts:100]: Hmm actually `undefined` is not a valid type in JSON, so `JSON.parse` will remove it

File: src/auth/auth.controller.ts (hunk 1 above)
Comment (saengowp) [reply src/auth/auth.controller.ts]: I think the best low-cost solution is to just parse the url, and verify that origin and path are in the allowlist.

Comment (saengowp) [reply src/auth/auth.controller.ts]: Maybe put this in the issue backlog. I can see that this bug is not introduced in this pr.

**Review by bombnp — APPROVED** [review]

File: src/auth/auth.controller.ts (hunk 1 above)
Comment (paphonb) [reply src/auth/auth.controller.ts]: > how do u propose we check for origin then?

See ce3c407

**Pull request merged**